File: backend/models/investment.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from bson import ObjectId
from .user import PyObjectId
from enum import Enum

class InvestmentType(str, Enum):
    CRYPTOCURRENCY = "cryptocurrency"
    CDB = "cdb"
    STOCKS = "stocks"
    MUTUAL_FUND = "mutual_fund"

class Investment(BaseModel):
    id: Optional[PyObjectId] = Field(default_factory=PyObjectId, alias="_id")
    user_id: PyObjectId
    investment_type: InvestmentType
    asset_name: str
    symbol: Optional[str] = None  # For crypto/stocks
    quantity: float
    purchase_price: float
    current_price: float
    total_invested: float
    current_value: float
    profit_loss: float
    profit_loss_percentage: float
    purchase_date: datetime
    maturity_date: Optional[datetime] = None  # For CDB
    interest_rate: Optional[float] = None  # For CDB
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    is_active: bool = True

    class Config:
        allow_population_by_field_name = True
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}

class CryptoCurrency(BaseModel):
    symbol: str
    name: str
    current_price: float
    price_change_24h: float
    price_change_percentage_24h: float
    market_cap: float
    volume_24h: float

class CDBOption(BaseModel):
    id: str
    name: str
    bank_name: str = "BankSys"
    interest_rate: float
    minimum_investment: float
    maturity_months: int
    type: str = "prefixed"  # prefixed, postfixed
    description: str

class InvestmentCreate(BaseModel):
    investment_type: InvestmentType
    asset_name: str
    symbol: Optional[str] = None
    quantity: float
    purchase_price: float

class InvestmentResponse(BaseModel):
    id: str
    investment_type: InvestmentType
    asset_name: str
    symbol: Optional[str] = None
    quantity: float
    purchase_price: float
    current_price: float
    current_value: float
    profit_loss: float
    profit_loss_percentage: float
    purchase_date: datetime

class PortfolioSummary(BaseModel):
    total_invested: float
    current_value: float
    total_profit_loss: float
    total_profit_loss_percentage: float
    investments_by_type: List[dict]

File: frontend/src/models/Investment.ts
export enum InvestmentType {
  CRYPTOCURRENCY = "cryptocurrency",
  CDB = "cdb",
  STOCKS = "stocks",
  MUTUAL_FUND = "mutual_fund"
}

export interface Investment {
  id: string;
  investment_type: InvestmentType;
  asset_name: string;
  symbol?: string;
  quantity: number;
  purchase_price: number;
  current_price: number;
  current_value: number;
  profit_loss: number;
  profit_loss_percentage: number;
  purchase_date: string;
}

export interface CreateInvestment {
  investment_type: InvestmentType;
  asset_name: string;
  symbol?: string;
  quantity: number;
  purchase_price: number;
}

export interface PortfolioSummary {
  total_invested: number;
  current_value: number;
  total_profit_loss: number;
  total_profit_loss_percentage: number;
  investments_by_type: Array<{
    type: string;
    invested: number;
    current_value: number;
    count: number;
    profit_loss: number;
    profit_loss_percentage: number;
  }>;
}

export interface CryptoCurrency {
  symbol: string;
  name: string;
  current_price: number;
  price_change_24h: number;
  price_change_percentage_24h: number;
  market_cap: number;
  volume_24h: number;
}

export interface CDBOption {
  id: string;
  name: string;
  bank_name: string;
  interest_rate: number;
  minimum_investment: number;
  maturity_months: number;
  type: string;
  description: string;
}